### PR TITLE
Subscription feature

### PR DESCRIPTION
**Pull request by tejaskh3**
Date: 8 oct, 2024

Developer Name: @tejaskh3 

---

## Issue Ticket Number:-

- [link](https://github.com/Real-Dev-Squad/website-backend/issues/2123)

## Description:
- this PR introduces the subscription feature, in which we are collecting user emails and phone numbers to send them notifications in the future.
- in the next PR we I'll add test cases to this feature. here is the link for next PR https://github.com/Real-Dev-Squad/website-www/pull/938/files

Is Under Feature Flag

- [x] Yes
- [ ] No

Database changes

- [x] Yes
- [] No

Breaking changes (If your feature is breaking/missing something please mention pending tickets)

- [ ] Yes
- [x] No

Is Development Tested?

- [x] Yes
- [ ] No

Tested in staging?

- [x] Yes
- [ ] No

### Add relevant Screenshot below ( e.g test coverage etc. )
[screen-capture (27).webm](https://github.com/user-attachments/assets/4a6be338-8dd2-461f-ab10-1a52b403dfd9)

### CODE REVIEW

File: app/controllers/subscribe.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  get isLoggedIn() {
+    return this.login.isLoggedIn;
+  }
+
+  get isSubscribed() {
+    return this.login.userData.isSubscribed;
+  }
+
+  get isSubmitDisabled() {
+    return !this.email || (this.phone && !/^\+91\d{10}$/.test(this.phone));
```
Comment (vinit717) [inline app/controllers/subscribe.js:32]: can you move the regex to constant

File: app/styles/subscribe.module.css [hunk 2]
```diff
@@ -0,0 +1,295 @@
+@import url("https://fonts.googleapis.com/css2?family=Inter:ital,opsz,wght@0,14..32,100..900;1,14..32,100..900&display=swap");
```
Comment (vinit717) [inline app/styles/subscribe.module.css:1]: Is this is align with the main site?

File: tests/unit/services/subcribe-test.js
Comment (vinit717) [inline tests/unit/services/subcribe-test.js:1]: What about other tests?

Comment (tejaskh3) [reply tests/unit/services/subcribe-test.js:1]: I have removed all the test cases from it, and adding it to the new PR.

File: app/controllers/subscribe.js (hunk 1 above)
Comment (tejaskh3) [reply app/controllers/subscribe.js:32]: sure!

File: app/styles/subscribe.module.css (hunk 2 above)
Comment (tejaskh3) [reply app/styles/subscribe.module.css:1]: Didn't get you. Please tell me

File: app/components/header.hbs [hunk 3]
```diff
@@ -51,6 +51,12 @@
           class="nav__element"
           href={{this.STATUS_URL}}
         >Status</a></li>
+        {{!-- {{#if @dev}} --}}
```
Comment (MehulKChaudhari) [inline app/components/header.hbs:54]: If not needed please remove this comments

Comment (tejaskh3) [reply app/components/header.hbs:54]: okay

File: app/controllers/subscribe.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  @action
+  updateEmail(event) {
+    this.email = event.target.value;
+  }
+
+  @action
+  updatePhone(event) {
+    const input = event.target.value;
+    this.phone = input.replace(/[^\d+]/g, '');
```
Comment (MehulKChaudhari) [inline app/controllers/subscribe.js:57]: If you have made a constants file, you can add this regex in there and then use it

File: app/routes/subscribe.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+
+export default class SubscribeRoute extends Route {
+  @service login;
```
Comment (MehulKChaudhari) [inline app/routes/subscribe.js:5]: If we are not using login service here, why its `injected`. 

you can remove it

File: app/controllers/subscribe.js (hunk 4 above)
Comment (tejaskh3) [reply app/controllers/subscribe.js:57]: Removed the comments

File: app/routes/subscribe.js (hunk 5 above)
Comment (tejaskh3) [reply app/routes/subscribe.js:5]: removed

**Review by MehulKChaudhari — DISMISSED** [review]

**Comment by MehulKChaudhari** [comment]
@tejaskh3 Please add tests of it

File: app/components/header.hbs [hunk 6]
```diff
@@ -51,6 +51,11 @@
           class="nav__element"
           href={{this.STATUS_URL}}
         >Status</a></li>
+        {{#if @dev}}
+        <li>
+          <LinkTo @route="subscribe" class="nav__element">Subscribe</LinkTo>
+        </li>
+      {{/if}}
```
Comment (MehulKChaudhari) [inline app/components/header.hbs:58]: Why this new `if` block? you can add it in below as both are `{{#if @dev}}`

File: app/components/header.hbs [hunk 7]
```diff
@@ -51,6 +51,11 @@
...
       {{#if @dev}}
```
Comment (MehulKChaudhari) [inline app/components/header.hbs:59]:
```suggestion
        {{#if @dev}}
        <li>
          <LinkTo @route="subscribe" class="nav__element">Subscribe</LinkTo>
        </li>
```

**Review by MehulKChaudhari — COMMENTED** [review]
Rest all the things looks good.

**Review by MehulKChaudhari — APPROVED** [review]

File: app/controllers/subscribe.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            phoneNumber: this.phone,
+          }),
+          headers: {
+            'Content-Type': 'application/json',
+          },
+          credentials: 'include',
+        });
+        if (!response.ok) {
+          this.toast.error(
+            'Something went wrong. Please try again.',
```
Comment (samarpan1738) [inline app/controllers/subscribe.js:80]: Can this error message be extracted in a constants file?

File: app/controllers/subscribe.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            '',
+            TOAST_OPTIONS,
+          );
+        } else {
+          this.login.userData.isSubscribed = true;
+          this.toggleSubscriptionModal();
+          this.toast.info('🎉 Thank you for subscribing!', '', TOAST_OPTIONS);
```
Comment (samarpan1738) [inline app/controllers/subscribe.js:87]: Can this success message be extracted in a constants file?

File: app/controllers/subscribe.js (hunk 8 above)
Comment (tejaskh3) [reply app/controllers/subscribe.js:80]: This is not reusable I check in the repo, and all the messages are passed are like this. So, can we please keep it here

File: app/controllers/subscribe.js (hunk 9 above)
Comment (tejaskh3) [reply app/controllers/subscribe.js:87]: This is not reusable I check in the repo, and all the messages are passed are like this. So, can we please keep it here

**Review by joyguptaa — APPROVED** [review]

File: app/controllers/subscribe.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    this.isFormOpen = !this.isFormOpen;
+  }
+
+  @action
+  toggleSubscriptionModal() {
+    this.showSubscriptionModal = !this.showSubscriptionModal;
+    if (!this.showSubscriptionModal) {
+      window.location.reload();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (tejaskh3) [reply app/controllers/subscribe.js]: Initially the states was loading when we close the modal. But I discussed this issue with @akankshadharkar he said we don't even need the modal here, just show the message on the page. And, now I have fixed this issue.

File: app/controllers/subscribe.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  get isLoggedIn() {
+    return this.login.isLoggedIn;
+  }
+
+  get isSubscribed() {
+    return this.login.userData.isSubscribed;
+  }
+
+  get isSubmitDisabled() {
+    return !this.email || (this.phone && !/^\+?\d*$/.test(this.phone));
```
Comment (tejaskh3) [inline app/controllers/subscribe.js:33]: Currently we are checking here. I have discussed this with @ankushdharkar we will add country code library with flag here. I have created flag here.
https://github.com/Real-Dev-Squad/website-backend/issues/2210

File: app/controllers/subscribe.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+  @action
+  updateEmail(event) {
+    this.email = event.target.value;
+  }
+
+  @action
+  updatePhone(event) {
+    const input = event.target.value;
+    const isValidPhone = /^\+?\d*$/.test(input);
```
Comment (tejaskh3) [inline app/controllers/subscribe.js:49]: Currently we are checking here. I have discussed this with @ankushdharkar we will add country code library with flag here. I have created flag here.
https://github.com/Real-Dev-Squad/website-backend/issues/2210

**Review by iamitprakash — DISMISSED** [review]
I do see contents are shifting up after submission, please fix this

**Comment by tejaskh3** [comment]
> I do see contents are shifting up after submission, please fix this

fixed this issue sir please check again.

**Comment by vinit717** [comment]
Can you update the video

**Comment by tejaskh3** [comment]
> Can you update the video

this is updated video sir, we have removed the modal.

**Review by vinit717 — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**